File: day01/cpp/main.cpp
#include <algorithm>
#include <fstream>
#include <functional>
#include <iostream>
#include <ranges>
#include <sstream>
#include <string>

using namespace std;

void q1(fstream &fileStream);
void q2(fstream &fileStream);

int main(int argc, char **argv) {
  if (argc != 2) {
    cout << "Usage: ./" << argv[0] << " <file_location>" << endl;
    return 1;
  }

  cout << argv[1] << endl;
  auto fileStream = fstream(argv[1]);

  if (!fileStream) {
    cout << "Couldn't open file " << argv[1] << endl;
    return 2;
  }

  q2(fileStream);

  return 0;
}

void q1(fstream &fileStream) {
  vector<int> left;
  vector<int> right;

  for (string line; getline(fileStream, line);) {
    stringstream ss(line);
    int l, r;

    ss >> l;
    ss >> r;

    left.push_back(l);
    right.push_back(r);
  }

  ranges::sort(left);
  ranges::sort(right);

  auto res =
      ranges::views::zip(left, right) | ranges::views::transform([](auto it) {
        return abs(get<0>(it) - get<1>(it));
      });

  cout << ranges::fold_left(res, 0, plus<int>{}) << endl;
}

void q2(fstream &fileStream) {
  vector<int> left;
  unordered_map<int, int> freq;

  for (string line; getline(fileStream, line);) {
    stringstream ss(line);
    int l, r;

    ss >> l;
    ss >> r;

    left.push_back(l);
    freq[r]++;
  }

  auto similarities =
      ranges::views::transform(left, [&](int l) { return l * freq[l]; });
  auto result = ranges::fold_left(similarities, 0, plus<int>{});

  cout << result << endl;
}

File: day01/rust/src/main.rs
use std::{
    collections::HashMap,
    fs::File,
    io::{BufRead, BufReader},
    process::exit,
};

fn main() {
    let mut args = std::env::args();
    let program_name = args.next().unwrap();

    if args.len() != 1 {
        println!("Usage: {} <file_location>", program_name);
        exit(1);
    }

    let file_location = args.next().unwrap();
    let file = File::open(&file_location);

    if file.is_err() {
        println!("Couldn't open file {}", file_location);
        exit(2);
    }

    let reader = BufReader::new(file.unwrap());

    q2(reader);
}

fn q1(reader: BufReader<File>) {
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();

    for line in reader.lines() {
        if let Ok(content) = line {
            let mut split = content.split_whitespace().map(|v| v.parse().unwrap());

            left.push(split.next().unwrap());
            right.push(split.next().unwrap());
        }
    }

    left.sort();
    right.sort();

    let diff: i32 = left
        .into_iter()
        .zip(right)
        .map(|it| (it.0 - it.1).abs())
        .sum();

    println!("{}", diff);
}

fn q2(reader: BufReader<File>) {
    let mut left: Vec<i32> = Vec::new();
    let mut freq: HashMap<i32, i32> = HashMap::new();

    for line in reader.lines() {
        if let Ok(content) = line {
            let mut split = content.split_whitespace().map(|v| v.parse().unwrap());

            left.push(split.next().unwrap());
            let right = split.next().unwrap();

            freq.insert(right, freq.get(&right).unwrap_or(&0) + 1);
        }
    }

    let diff: i32 = left
        .into_iter()
        .map(|it| (it * freq.get(&it).unwrap_or(&0)))
        .sum();

    println!("{}", diff);
}
